rfcs/0134-nix-store-layer.md

```diff
@@ -0,0 +1,260 @@
+There are a few tweaks and generalizations we can make, but I struggle to envision what wildly different alternatives one might want instead.
+
+A small, stable interface that allows for design exploration above and below it is known as a [narrow waist](https://www.oilshell.org/cross-ref.html?tag=narrow-waist#narrow-waist).
+It's not every day that a project happens upon a great narrow waist design.
+I believe we've discovered a very good one with Nix, and that should be seen as a *key asset*, even if it is not how we recruit "regular users".
+
+By making a store-only Nix, we put more emphasis on this key interface.
+All functionality the store-only Nix offers factors through this interface.
+The upper half of Nix likewise uses the lower half through this interface.
+The daemon protocol represents this interface for IPC, and allows either half to be swapped for a different implementation.
+
+To help explain the community-building benefits, it might help to go over some specific examples.
+
+### Tvix and go-nix
+
+[TVL announced](https://tvl.fyi/blog/rewriting-nix) that, frustrated in trying to refactor Nix into something more modular and flexible, they were aiming to make a new implementation from scratch.
+More recently, [they lay out a basic approach](https://tvl.fyi/blog/tvix-status-september-22) of two projects:
+
+- [Tvix](https://cs.tvl.fyi/depot/-/tree/tvix) is a new implementation of the Nix language evaluator,
+
+- [go-nix](https://github.com/nix-community/go-nix) is a new implementation of the store layer.
+
+First of all, the fact that they are planning on two completely separate implementation oriented around this same "narrow waist" is testament to the appeal of the design.
+
+Second of all, note that they have separate, orthogonal experiments they wish to run on both sides of the store interface divide.
+Above, they want to experiment with radically different evaluation strategies, especially to speed up Nixpkgs evaluation.
+Below, they want to experiment with the standardized containerization technologies that already exist for new ways of sandboxing and distributing builds with less bespoke Nix-specific code.
+They also want to apply the layering paradigm *within* go-nix, fostering even more modularity.
+
+I think these are great goals, and for the sake of the ecosystem as a whole, it should be as easy as possible to run such new experiments.
+In particular, a novel evaluator should be usable with the standard C++ Nix store layer, and a novel store layer should be reusable with the standard C++ Nix evaluator.
+
+Yes, strictly speaking, we only need a stable daemon protocol to accommodate that goal, which we have.
+But ensuring the C++ Nix components can be built separately for use in isolation further send the message that such experimentation is *good* and *welcomed*.
+Messaging matters, and making our layered architecture "official" as this RFC proposes I think sends a better message.
+
+### Guix
```

Comment:
Guix is not conceptional compatible on the store layer level with Nix. They support having scheme expression that are executed at build-time in their derivation files.

Reply:
Depends how they've integrated that, how that scheme is defined and whether that's inconsistent with any `system` value we've used in Nix.
I couldn't find anything obvious about this in the Guix manual. Do you have a link to this info?

Reply:
https://guix.gnu.org/manual/en/html_node/G_002dExpressions.html

Reply:
@Mic92 No that is just them swapping out scheme for bash.

https://guix.gnu.org/en/blog/2023/dissecting-guix-part-1-derivations/ is not entirely clear, but shows that the drv format is not affected by g-expressions.